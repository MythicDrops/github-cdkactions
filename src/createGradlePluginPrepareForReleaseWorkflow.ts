import { Job, Stack, Workflow } from "cdkactions";
import dedent from "ts-dedent";
import { GradlePluginConfig } from "./types";

export const createGradlePluginPrepareForReleaseWorkflow = (
  stack: Stack,
  config: GradlePluginConfig
): Workflow => {
  const workflow = new Workflow(stack, "prepare-for-release", {
    name: "Prepare For Release",
    on: {
      push: { branches: ["main"] },
    },
  });

  const pluginName = config.pluginName;
  const primaryGradleVersion = config.primaryGradleVersion ?? "7.0.2";
  const timeoutMinutes = config?.timeoutMinutes ?? 10;

  new Job(workflow, "prepare-for-release", {
    name: "Prepare For Release",
    runsOn: "ubuntu-latest",
    timeoutMinutes,
    steps: [
      {
        name: "Check-out Code",
        uses: "actions/checkout@v2",
        with: { "fetch-depth": 0 },
      },
      {
        name: "Turnstyle",
        uses: "softprops/turnstyle@v1",
        env: {
          GITHUB_TOKEN: `\${{ secrets.GITHUB_TOKEN }}`,
        },
      },
      {
        name: "Install GitVersion",
        uses: "gittools/actions/gitversion/setup@v0.9.9",
        with: { versionSpec: "5.x" },
      },
      {
        name: "Determine Version",
        id: "gitversion",
        uses: "gittools/actions/gitversion/execute@v0.9.9",
      },
      {
        name: "Set up JDK",
        uses: "actions/setup-java@v2",
        with: {
          distribution: "adopt",
          "java-version": "1.8",
        },
      },
      {
        name: "Make Gradle Wrapper Executable",
        run: dedent`chmod +x ./gradlew`,
      },
      {
        name: "Set Gradle Version",
        run: dedent`./gradlew wrapper --gradle-version ${primaryGradleVersion}`,
      },
      {
        name: "Test Plugin Directly with Gradle",
        run: dedent`../gradlew check`,
        workingDirectory: pluginName,
      },
      {
        name: "Test Plugin Application with Gradle",
        run: dedent`./gradlew check`,
      },
      {
        name: "Push Tag",
        uses: "mathieudutour/github-tag-action@v5.1",
        with: {
          custom_tag: `\${{ steps.gitversion.outputs.semVer }}`,
          github_token: `\${{ secrets.ACTIONS_PAT }}`,
          tag_prefix: "",
          create_annotated_tag: true,
        },
      },
      {
        name: "Publish Plugins to Gradle Plugin Portal",
        run: dedent`./gradlew :${pluginName}:publishPlugins \
          -Pversion=\${{ steps.gitversion.outputs.semVer }} \
          -Pgradle.publish.key=\${{ secrets.GRADLE_PUBLISH_KEY }} \
          -Pgradle.publish.secret=\${{ secrets.GRADLE_PUBLISH_SECRET }}`,
      },
    ],
  });
  return workflow;
};
