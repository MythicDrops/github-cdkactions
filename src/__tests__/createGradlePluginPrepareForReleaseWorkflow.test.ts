import { App, Stack } from "cdkactions";
import { createGradlePluginPrepareForReleaseWorkflow } from "../createGradlePluginPrepareForReleaseWorkflow";

describe("createGradlePluginPrepareForReleaseWorkflow", () => {
  let app: App;
  let stack: Stack;

  beforeEach(() => {
    app = new App();
    stack = new Stack(app, "cdk");
  });

  it("generates a GitHub Actions workflow", () => {
    const pluginName = "pixeloutlaw-gradle-plugin";
    expect(
      createGradlePluginPrepareForReleaseWorkflow(stack, {
        pluginName,
      }).toGHAction()
    ).toMatchSnapshot();
  });
});
