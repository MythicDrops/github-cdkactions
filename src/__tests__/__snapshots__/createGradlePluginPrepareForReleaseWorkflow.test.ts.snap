// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGradlePluginPrepareForReleaseWorkflow generates a GitHub Actions workflow 1`] = `
Object {
  "jobs": Object {
    "prepare-for-release": Object {
      "name": "Prepare For Release",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "name": "Check-out Code",
          "uses": "actions/checkout@v2",
          "with": Object {
            "fetch-depth": 0,
          },
        },
        Object {
          "env": Object {
            "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
          },
          "name": "Turnstyle",
          "uses": "softprops/turnstyle@v1",
        },
        Object {
          "name": "Install GitVersion",
          "uses": "gittools/actions/gitversion/setup@v0.9.9",
          "with": Object {
            "versionSpec": "5.x",
          },
        },
        Object {
          "id": "gitversion",
          "name": "Determine Version",
          "uses": "gittools/actions/gitversion/execute@v0.9.9",
        },
        Object {
          "name": "Set up JDK",
          "uses": "actions/setup-java@v2",
          "with": Object {
            "distribution": "adopt",
            "java-version": "1.8",
          },
        },
        Object {
          "name": "Make Gradle Wrapper Executable",
          "run": "chmod +x ./gradlew",
        },
        Object {
          "name": "Set Gradle Version",
          "run": "./gradlew wrapper --gradle-version 7.0.2",
        },
        Object {
          "name": "Test Plugin Directly with Gradle",
          "run": "../gradlew check",
          "working-directory": "pixeloutlaw-gradle-plugin",
        },
        Object {
          "name": "Test Plugin Application with Gradle",
          "run": "./gradlew check",
        },
        Object {
          "name": "Push Tag",
          "uses": "mathieudutour/github-tag-action@v5.1",
          "with": Object {
            "create_annotated_tag": true,
            "custom_tag": "\${{ steps.gitversion.outputs.semVer }}",
            "github_token": "\${{ secrets.ACTIONS_PAT }}",
            "tag_prefix": "",
          },
        },
        Object {
          "name": "Publish Plugins to Gradle Plugin Portal",
          "run": "./gradlew :pixeloutlaw-gradle-plugin:publishPlugins           -Pversion=\${{ steps.gitversion.outputs.semVer }}           -Pgradle.publish.key=\${{ secrets.GRADLE_PUBLISH_KEY }}           -Pgradle.publish.secret=\${{ secrets.GRADLE_PUBLISH_SECRET }}",
        },
      ],
      "timeout-minutes": 10,
    },
  },
  "name": "Prepare For Release",
  "on": Object {
    "push": Object {
      "branches": Array [
        "main",
      ],
    },
  },
}
`;
